File: src/core/marketUid.ts
import { createHash } from 'node:crypto';
const MARKET_UID_PREFIX = 'muid';
const MARKET_UID_VERSION = 'v1';

export interface MarketIdentifier {
  operator: string;
  sport: string;
  league: string;
  event: string;
  marketType: string;
  outcome: string;
  eventTimestamp: Date;
  variant?: string | null;
  ladder?: string | null;
}

export type MarketUid = string;
export class MarketUidError extends Error {
  readonly field: keyof MarketIdentifier;

  constructor(field: keyof MarketIdentifier) {
    super(`missing required field \`${String(field)}\` for market UID generation`);
    this.name = 'MarketUidError';
    this.field = field;
  }
}

export function canonicalFingerprint(identifier: MarketIdentifier): string {
  const operator = normalizeRequired(identifier.operator, 'operator');
  const sport = normalizeRequired(identifier.sport, 'sport');
  const league = normalizeRequired(identifier.league, 'league');
  const event = normalizeRequired(identifier.event, 'event');
  const marketType = normalizeRequired(identifier.marketType, 'marketType');
  const outcome = normalizeRequired(identifier.outcome, 'outcome');
  const variant = normalizeOptional(identifier.variant);
  const ladder = normalizeOptional(identifier.ladder);
  const eventTime = truncateTimestamp(identifier.eventTimestamp);

  return [
    MARKET_UID_PREFIX,
    MARKET_UID_VERSION,
    operator,
    sport,
    league,
    event,
    marketType,
    variant,
    ladder,
    eventTime,
    outcome,
  ].join('|');
}

export function marketUidFromIdentifier(identifier: MarketIdentifier): MarketUid {
  const fingerprint = canonicalFingerprint(identifier);
  const digest = createHash('sha256').update(fingerprint).digest('hex');
  return `${MARKET_UID_PREFIX}-${MARKET_UID_VERSION}-${digest.slice(0, 24)}`;
}

function normalizeRequired(
  value: string,
  field: keyof MarketIdentifier,
): string {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new MarketUidError(field);
  }

  const normalized = canonicalize(trimmed);
  if (!normalized) {
    throw new MarketUidError(field);
  }

  return normalized;
}

function normalizeOptional(value: string | null | undefined): string {
  if (value === undefined || value === null) {
    return 'na';
  }
  const trimmed = value.trim();
  if (!trimmed) {
    return 'na';
  }
  const normalized = canonicalize(trimmed);
  return normalized || 'na';
}

function truncateTimestamp(timestamp: Date): string {
  const millis = timestamp.getTime();
  const truncated = new Date(Math.trunc(millis / 60_000) * 60_000);
  const year = truncated.getUTCFullYear().toString().padStart(4, '0');
  const month = (truncated.getUTCMonth() + 1).toString().padStart(2, '0');
  const day = truncated.getUTCDate().toString().padStart(2, '0');
  const hours = truncated.getUTCHours().toString().padStart(2, '0');
  const minutes = truncated.getUTCMinutes().toString().padStart(2, '0');
  return `${year}${month}${day}T${hours}${minutes}Z`;
}

function canonicalize(value: string): string {
  let buffer = '';
  for (const char of value.normalize('NFKD')) {
    buffer += isAsciiAlphaNumeric(char) ? char.toLowerCase() : ' ';
  }
  return buffer.split(/\s+/).filter(Boolean).join('_');
}

function isAsciiAlphaNumeric(char: string): boolean {
  if (char.length !== 1) {
    return false;
  }

  const code = char.charCodeAt(0);
  return (
    (code >= 48 && code <= 57) ||
    (code >= 65 && code <= 90) ||
    (code >= 97 && code <= 122)
  );
}

File: crates/normalization/src/market_uid.rs
use chrono::{DateTime, Timelike, Utc};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::fmt;
use thiserror::Error;

const MARKET_UID_PREFIX: &str = "muid";
const MARKET_UID_VERSION: &str = "v1";

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct MarketIdentifier {
    pub operator: String, pub sport: String, pub league: String, pub event: String,
    pub market_type: String, pub outcome: String, pub event_timestamp: DateTime<Utc>,
    pub variant: Option<String>, pub ladder: Option<String>,
}

impl MarketIdentifier {
    pub fn canonical_fingerprint(&self) -> Result<String, MarketUidError> {
        let operator = normalize_required(&self.operator, "operator")?;
        let sport = normalize_required(&self.sport, "sport")?;
        let league = normalize_required(&self.league, "league")?;
        let event = normalize_required(&self.event, "event")?;
        let market_type = normalize_required(&self.market_type, "market_type")?;
        let outcome = normalize_required(&self.outcome, "outcome")?;
        let ladder = normalize_optional(self.ladder.as_deref());
        let variant = normalize_optional(self.variant.as_deref());
        let event_time = truncate_timestamp(self.event_timestamp);

        Ok(format!(
            "{MARKET_UID_PREFIX}|{MARKET_UID_VERSION}|{operator}|{sport}|{league}|{event}|{market_type}|{variant}|{ladder}|{event_time}|{outcome}"
        ))
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct MarketUid(String);

impl MarketUid {
    pub fn from_identifier(identifier: &MarketIdentifier) -> Result<Self, MarketUidError> {
        let fingerprint = identifier.canonical_fingerprint()?;
        let mut hasher = Sha256::new();
        hasher.update(fingerprint.as_bytes());
        let digest = hasher.finalize();
        let hash = hex::encode(digest);
        let value = format!("{MARKET_UID_PREFIX}-{MARKET_UID_VERSION}-{}", &hash[..24]);
        Ok(Self(value))
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for MarketUid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

#[derive(Debug, Error)]
pub enum MarketUidError {
    #[error("missing required field `{0}` for market UID generation")]
    MissingField(&'static str),
}

fn normalize_required(value: &str, field: &'static str) -> Result<String, MarketUidError> {
    let cleaned = value.trim();
    if cleaned.is_empty() {
        Err(MarketUidError::MissingField(field))
    } else {
        Ok(canonicalize(cleaned))
    }
}

fn normalize_optional(value: Option<&str>) -> String {
    value
        .map(|v| canonicalize(v.trim()))
        .filter(|v| !v.is_empty())
        .unwrap_or_else(|| "na".to_string())
}

fn truncate_timestamp(timestamp: DateTime<Utc>) -> String {
    timestamp
        .with_second(0)
        .and_then(|dt| dt.with_nanosecond(0))
        .unwrap_or(timestamp)
        .format("%Y%m%dT%H%MZ")
        .to_string()
}

fn canonicalize(input: &str) -> String {
    let normalized: String = input
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c.to_ascii_lowercase() } else { ' ' })
        .collect();
    normalized
        .split_whitespace()
        .map(str::to_string)
        .collect::<Vec<_>>()
        .join("_")
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    #[test]
    fn market_uid_generation_behaves() {
        let identifier = MarketIdentifier { operator: "sx".into(), sport: "Soccer".into(), league: "Premier League".into(), event: "Arsenal vs Chelsea".into(), market_type: "moneyline".into(), outcome: "home".into(), event_timestamp: Utc.with_ymd_and_hms(2024, 5, 1, 18, 30, 45).unwrap(), variant: Some("pre".into()), ladder: None };
        let uid = MarketUid::from_identifier(&identifier).unwrap();
        assert!(uid.as_str().starts_with("muid-v1-"));

        let missing_operator = MarketIdentifier { operator: "".into(), sport: "Soccer".into(), league: "Premier League".into(), event: "Arsenal vs Chelsea".into(), market_type: "moneyline".into(), outcome: "home".into(), event_timestamp: Utc::now(), variant: None, ladder: None };
        let error = MarketUid::from_identifier(&missing_operator).unwrap_err();
        assert!(matches!(error, MarketUidError::MissingField("operator")));
    }
}
